test/junit/scala/collection/immutable/MapHashcodeTest.scala

```diff
@@ -0,0 +1,292 @@
+    val t1 = new Map.Map1(1, 1)
+    val t2 = new Map.Map2(1, 1, 2, 2)
+    val t3 = new Map.Map3(1, 1, 2, 2, 3, 3)
+    val t4 = new Map.Map4(1, 1, 2, 2, 3, 3, 4, 4)
+
+    nonAllocating {
+      t0.hashCode()
+    }
+    nonAllocating {
+      t0.##
+    }
+
+    nonAllocating {
+      t1.hashCode()
+    }
+    nonAllocating {
+      t1.##
+    }
+
+    nonAllocating {
+      t2.hashCode()
+    }
+    nonAllocating {
+      t2.##
+    }
+
+    nonAllocating {
+      t3.hashCode()
+    }
+    nonAllocating {
+      t3.##
+    }
+
+    nonAllocating {
+      t4.hashCode()
+    }
+    nonAllocating {
+      t4.##
+    }
+  }
+
+  @Test def nonAllocatingListMap(): Unit = {
+    val t0 = ListMap.empty[String, String]
+    val t1 = t0.updated("1", "1")
+    val t2 = t1.updated("2", "2")
+    val t3 = t2.updated("3", "3")
+    val t4 = t3.updated("4", "4")
+
+    val tlarge = (1 to 10000).foldLeft(t0) {
+      (a, b) => a.updated(b.toString, b.toString)
+    }
+
+    nonAllocating {
+      t0.hashCode()
+    }
+    nonAllocating {
+      t0.##
+    }
+
+    onlyAllocates(32) {
+      t1.hashCode()
+    }
+    onlyAllocates(32) {
+      t1.##
+    }
+
+    onlyAllocates(32) {
+      t2.hashCode()
+    }
+    onlyAllocates(32) {
+      t2.##
+    }
+
+    onlyAllocates(32) {
+      t3.hashCode()
+    }
+    onlyAllocates(32) {
+      t3.##
+    }
+
+    onlyAllocates(32) {
+      t4.hashCode()
+    }
+    onlyAllocates(32) {
+      t4.##
+    }
+
+    onlyAllocates(32) {
+      tlarge.hashCode()
+    }
+    onlyAllocates(32) {
+      tlarge.##
+    }
+  }
+
+  @Test def nonAllocatingSortedMap(): Unit = {
+    val t0 = SortedMap.empty[String, String]
+    val t1 = t0.updated("1", "1")
+    val t2 = t1.updated("2", "2")
+    val t3 = t2.updated("3", "3")
+    val t4 = t3.updated("4", "4")
+
+    val tlarge = (1 to 10000).foldLeft(t0) {
+      (a, b) => a.updated(b.toString, b.toString)
+    }
+
+    onlyAllocates(32) {
+      t0.hashCode()
+    }
+    onlyAllocates(32) {
+      t0.##
+    }
+
+    onlyAllocates(32) {
+      t1.hashCode()
+    }
+    onlyAllocates(32) {
+      t1.##
+    }
+
+    onlyAllocates(32) {
+      t2.hashCode()
+    }
+    onlyAllocates(32) {
+      t2.##
+    }
+
+    onlyAllocates(32) {
+      t3.hashCode()
+    }
+    onlyAllocates(32) {
+      t3.##
+    }
+
+    onlyAllocates(32) {
+      t4.hashCode()
+    }
+    onlyAllocates(32) {
+      t4.##
+    }
+
+    onlyAllocates(32) {
+      tlarge.hashCode()
+    }
+    onlyAllocates(32) {
+      tlarge.##
+    }
+  }
+
+  @Test def nonAllocatingHashMap(): Unit = {
+    val t0 = HashMap.empty[String, String]
+    val t1 = t0.updated("1", "1")
+    val t2 = t1.updated("2", "2")
+    val t3 = t2.updated("3", "3")
+    val t4 = t3.updated("4", "4")
+
+    val tlarge = (1 to 10000).foldLeft(t0) {
+      (a, b) => a.updated(b.toString, b.toString)
+    }
+
+    nonAllocating {
+      t0.hashCode()
+    }
+    nonAllocating {
+      t0.##
+    }
+
+    onlyAllocates(256) {
```

Comment:
This is higher than in the 2.12.x version because we use [`ChampBaseIterator`](https://github.com/scala/scala/blob/d78703a420cd9bc43396107d875363e60c910f35/src/library/scala/collection/immutable/ChampCommon.scala#L100-L187) which avoids using the JVM stack during iteration into the `Trie`.